sql-plugin/src/main/scala/com/nvidia/spark/rapids/literals.scala

```diff
@@ -79,7 +91,132 @@ object GpuScalar {
+   *
+   * It does not handle null because null is filtered out in 'from'.
+   *
+   * NOTE: This is done by leveraging the `ColumnVector.fromXXX` API family. These APIs are
+   * designed for test only according to its doc, but it is ok to use here, since the size
+   * of the literal values will not be large in most cases.
+   *
+   * @param data the data to build the Scalar. Should not be null.
+   * @param elementType the element type
+   * @return a cudf Scalar contains the array data.
+   */
+  private def createScalarFromArray(data: ArrayData, elementType: DataType): Scalar = {
+    val array = data.array
+    val listView = elementType match {
+      // Uses the boxed version for primitive types to keep the nulls.
+      case ByteType => ColumnVector.fromBoxedBytes(castTo[JByte](array): _*)
+      case DateType => ColumnVector.timestampDaysFromBoxedInts(castTo[Integer](array): _*)
+      case LongType => ColumnVector.fromBoxedLongs(castTo[JLong](array): _*)
+      case ShortType => ColumnVector.fromBoxedShorts(castTo[JShort](array): _*)
+      case FloatType => ColumnVector.fromBoxedFloats(castTo[JFloat](array): _*)
+      case DoubleType => ColumnVector.fromBoxedDoubles(castTo[JDouble](array): _*)
+      case IntegerType => ColumnVector.fromBoxedInts(castTo[Integer](array): _*)
+      case BooleanType => ColumnVector.fromBoxedBooleans(castTo[JBoolean](array): _*)
+      case TimestampType =>
+        ColumnVector.timestampMicroSecondsFromBoxedLongs(castTo[JLong](array): _*)
+      case StringType =>
+        ColumnVector.fromStrings(castTo[UTF8String, String](array)(convertUTF8StringTo(_)): _*)
+      case dt: DecimalType =>
+        if (DecimalType.is32BitDecimalType(dt)) {
+          val rows = castTo[Decimal, Integer](array)(convertDecimalTo(_, dt).intValue())
+          ColumnVector.decimalFromBoxedInts(-dt.scale, rows: _*)
+        } else {
+          val rows = castTo[Decimal, JLong](array)(convertDecimalTo(_, dt))
+          ColumnVector.decimalFromBoxedLongs(-dt.scale, rows: _*)
+        }
+      case ArrayType(_, _) =>
+        val colType = resolveElementType(elementType)
+        val rows = castTo[JList[_]](convertElementTo(_, elementType))(array)
+        ColumnVector.fromLists(colType, rows: _*)
+      case StructType(_) =>
+        val colType = resolveElementType(elementType)
+        val rows = castTo[HostColumnVector.StructData](convertElementTo(_, elementType))(array)
+        ColumnVector.fromStructs(colType, rows: _*)
+      case u =>
+        throw new IllegalStateException(s"Unsupported element type ($u) for the list scalar.")
+    }
+
+    withResource(listView) { list =>
+      Scalar.listFromColumnView(list);
+    }
+  }
+
+  @deprecated("This will be removed. Since no need to infer the type again because" +
+    " Spark does it already. Besides it is difficult to support nested type.", "")
   def from(v: Any): Scalar = v match {
+    case s: Scalar => s.incRefCount()
```

Comment:
If `GpuScalar.from` is going to handle `Scalar` directly, is it still required to explicitly handle `Scalar` after an `expr.columnarEval`?  Seems like we can let the existing `case other => GpuScalar.from...` code expand the scalar to a column.

Reply:
This is to cover the case that some nodes will close a closeable object returned from `expr.columnarEval`. e.g. `GpuBinaryExpression` and `GpuUnaryExpression` .

With this, no need to add the `scalar ` case specially and the `scalar` is treated as a closeable object.

Reply:
We may want to unify the result handling of `expr.columnarEval` for these two kinds of cases.

Reply:
This can also be trakced by the issue #2327